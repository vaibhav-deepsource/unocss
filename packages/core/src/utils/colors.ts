import type { RGBAColorValue } from '../types'

/* eslint-disable no-case-declarations */
const hexRE = /^#?([\da-f]+)$/i

export function hex2rgba(hex = ''): RGBAColorValue | undefined {
  const [, body] = hex.match(hexRE) || []

  if (!body)
    return

  switch (body.length) {
    case 3:
    case 4:
      const digits = Array.from(body, s => Number.parseInt(s, 16)).map(n => (n << 4) | n)
      if (body.length === 3)
        return digits as [number, number, number]
      digits[3] = Math.round(digits[3] / 255 * 100) / 100
      return digits as [number, number, number, number]
    case 6:
    case 8:
      const value = Number.parseInt(body, 16)
      if (body.length === 6)
        return [(value >> 16) & 0xFF, (value >> 8) & 0xFF, value & 0xFF]
      return [(value >> 24) & 0xFF, (value >> 16) & 0xFF, (value >> 8) & 0xFF, Math.round((value & 0xFF) / 255 * 100) / 100]
  }
}
