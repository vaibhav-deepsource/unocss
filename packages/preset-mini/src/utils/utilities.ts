import type { CSSEntries, CSSObject, DynamicMatcher, ParsedColorValue, RuleContext } from '@unocss/core'
import { hex2rgba } from '@unocss/core'
import type { Theme } from '../theme'
import { handler as h } from './handlers'
import { directionMap } from './mappings'

/**
 * Provide {@link DynamicMatcher} function returning spacing definition. See spacing rules.
 *
 * @param {string} propertyPrefix - Property for the css value to be created. Postfix will be appended according to direction matched.
 * @return {DynamicMatcher}  {@link DynamicMatcher}
 * @see {@link directionMap}
 */
export const directionSize = (propertyPrefix: string): DynamicMatcher => ([_, direction, size]: string[]): CSSEntries | undefined => {
  const v = h.bracket.cssvar.auto.fraction.rem(size)
  if (v !== undefined)
    return directionMap[direction].map(i => [`${propertyPrefix}${i}`, v])
}

/**
 * Obtain color from theme by camel-casing colors.
 */
const getThemeColor = (theme: Theme, colors: string[]) =>
  theme.colors?.[
    colors.join('-').replace(/(-[a-z])/g, n => n.slice(1).toUpperCase())
  ]

/**
 * Parse color string into rgba (if possible) with opacity. Color value will be matched to theme object before converting to rgb value.
 *
 * @example Parseable strings:
 * 'red' // From theme, if 'red' is available
 * 'red-100' // From theme, plus scale
 * 'red-100/20' // From theme, plus scale/opacity
 * '#f12' // Hex color
 * 'hex-f12' // Alternative hex color
 * '[rgb(100,2,3)]/[var(--op)]' // Bracket with rgb color and bracket with opacity
 *
 * @param {string} body - Color string to be parsed.
 * @param {Theme} theme - {@link Theme} object.
 * @return {ParsedColorValue|undefined}  {@link ParsedColorValue} object if string is parseable.
 */
export const parseColor = (body: string, theme: Theme): ParsedColorValue | undefined => {
  const [main, opacity] = body.split(/(?:\/|:)/)
  const colors = main
    .replace(/([a-z])([0-9])/g, '$1-$2')
    .split(/-/g)
  const [name] = colors

  if (!name)
    return

  let color: string | undefined
  const bracket = h.bracket(main)
  const bracketOrMain = bracket || main

  if (bracketOrMain.startsWith('#'))
    color = bracketOrMain.slice(1)
  else if (bracketOrMain.startsWith('hex-'))
    color = bracketOrMain.slice(4)
  else if (main.startsWith('$'))
    color = h.cssvar(main)

  color = color || bracket

  let no = 'DEFAULT'
  if (!color) {
    let colorData
    const [scale] = colors.slice(-1)
    if (scale.match(/^\d+$/)) {
      no = scale
      colorData = getThemeColor(theme, colors.slice(0, -1))
    }
    else {
      colorData = getThemeColor(theme, colors)
      if (!colorData) {
        [, no = no] = colors
        colorData = getThemeColor(theme, [name])
      }
    }

    if (typeof colorData === 'string')
      color = colorData
    else if (no && colorData)
      color = colorData[no]
  }

  const rgba = hex2rgba(color)

  const alpha = opacity
    ? opacity[0] === '['
      ? h.bracket.percent(opacity)!
      : (parseFloat(opacity) / 100)
    : rgba?.[3]

  const hasAlpha = alpha != null && !Number.isNaN(alpha)
  if (rgba) {
    if (hasAlpha) {
      rgba[3] = typeof alpha === 'string' && !alpha.includes('%')
        ? parseFloat(alpha)
        : alpha as number
    }
    else {
      rgba.splice(3)
    }
  }

  return {
    opacity,
    name,
    no,
    color,
    rgba,
    alpha: hasAlpha ? alpha : undefined,
  }
}

/**
 * Provide {@link DynamicMatcher} function to produce color value matched from rule.
 *
 * @see {@link parseColor}
 *
 * @example Resolving 'red' from theme:
 * colorResolver('background-color', 'background')('', 'red')
 * return { 'background-color': '#f12' }
 *
 * @example Resolving 'red-100' from theme:
 * colorResolver('background-color', 'background')('', 'red-100')
 * return { '--un-background-opacity': '1', 'background-color': 'rgba(254,226,226,var(--un-bg-opacity))' }
 *
 * @example Resolving 'red-100/20' from theme:
 * colorResolver('background-color', 'background')('', 'red-100/20')
 * return { 'background-color': 'rgba(204,251,241,0.22)' }
 *
 * @example Resolving 'hex-124':
 * colorResolver('color', 'text')('', 'hex-124')
 * return { '--un-text-opacity': '1', 'color': 'rgba(17,34,68,var(--un-text-opacity))' }
 *
 * @param {string} property - Property for the css value to be created.
 * @param {string} varName - Base name for the opacity variable.
 * @return {DynamicMatcher}  {@link DynamicMatcher} object.
 */
export const colorResolver = (property: string, varName: string): DynamicMatcher => ([, body]: string[], { theme }: RuleContext<Theme>): CSSObject | undefined => {
  const data = parseColor(body, theme)

  if (!data)
    return

  const { alpha, opacity, color, rgba } = data

  if (!color)
    return

  if (rgba) {
    if (alpha != null) {
      return {
        [property]: `rgba(${rgba.join(',')})`,
      }
    }
    else {
      return {
        [`--un-${varName}-opacity`]: (opacity && h.cssvar(opacity)) ?? 1,
        [property]: `rgba(${rgba.join(',')},var(--un-${varName}-opacity))`,
      }
    }
  }
  else {
    return {
      [property]: color.replace('%alpha', `${alpha || 1}`),
    }
  }
}
