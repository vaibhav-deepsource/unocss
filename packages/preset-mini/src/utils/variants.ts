import type { VariantHandler } from '@unocss/core'
import { escapeRegExp } from '@unocss/core'

export const variantMatcher = (name: string, selector?: (input: string) => string | undefined) => {
  const re = new RegExp(`^${escapeRegExp(name)}[:-]`)
  return (input: string): VariantHandler | undefined => {
    const match = input.match(re)
    if (match) {
      return {
        matcher: input.slice(match[0].length),
        selector,
      }
    }
  }
}

export const variantParentMatcher = (name: string, parent: string) => {
  const re = new RegExp(`^${escapeRegExp(name)}[:-]`)
  return (input: string): VariantHandler | undefined => {
    const match = input.match(re)
    if (match) {
      return {
        matcher: input.slice(match[0].length),
        parent,
      }
    }
  }
}
