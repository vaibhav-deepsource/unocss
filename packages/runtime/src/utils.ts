import type { Postprocessor } from '@unocss/core'

const camelize = (str: string) => str.replace(/-(\w)/g, (_, c) => c ? c.toUpperCase() : '')
const capitalize = (str: string) => str.charAt(0).toUpperCase() + str.slice(1)
const hyphenate = (str: string) => str.replace(/(?:^|\B)([A-Z])/g, '-$1').toLowerCase()
const prefixes = ['Webkit', 'Moz', 'ms']

export function autoPrefixer(style: CSSStyleDeclaration): Postprocessor {
  const prefixCache: Record<string, string> = {}

  function autoPrefix(rawName: string): string {
    const cached = prefixCache[rawName]
    if (cached)
      return cached
    let name = camelize(rawName)
    if (name !== 'filter' && name in style)
      return (prefixCache[rawName] = hyphenate(name))
    name = capitalize(name)
    for (let i = 0; i < prefixes.length; i++) {
      const prefixed = `${prefixes[i]}${name}`
      if (prefixed in style)
        return (prefixCache[rawName] = hyphenate(capitalize(prefixed)))
    }
    return rawName
  }

  return ({ entries }) => entries.forEach((e) => {
    if (!e[0].startsWith('--'))
      e[0] = autoPrefix(e[0])
  })
}
