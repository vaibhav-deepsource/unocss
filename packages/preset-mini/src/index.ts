import type { Postprocessor, Preset, PresetOptions } from '@unocss/core'
import { rules } from './rules'
import type { Theme, ThemeAnimation } from './theme'
import { theme } from './theme'
import { variants } from './variants'

export { theme, colors } from './theme'
export { parseColor } from './utils'

export type { ThemeAnimation, Theme }

export interface PresetMiniOptions extends PresetOptions {
  /**
   * @default 'class'
   */
  dark?: 'class' | 'media'
  /**
   * @default false
   */
  attributifyPseudo?: Boolean
  /**
   * @default 'un-'
   */
  variablePrefix?: string
}

export const presetMini = (options: PresetMiniOptions = {}): Preset<Theme> => {
  options.dark = options.dark ?? 'class'
  options.attributifyPseudo = options.attributifyPseudo ?? false

  return {
    name: '@unocss/preset-mini',
    theme,
    rules,
    variants: variants(options),
    options,
    postprocess: options.variablePrefix && options.variablePrefix !== 'un-'
      ? VarPrefixPostprocessor(options.variablePrefix)
      : undefined,
  }
}

export default presetMini

function VarPrefixPostprocessor(prefix: string): Postprocessor {
  return (obj) => {
    obj.entries.forEach((i) => {
      i[0] = i[0].replace(/^--un-/, `--${prefix}`)
      if (typeof i[1] === 'string')
        i[1] = i[1].replace(/var\(--un-/g, `var(--${prefix}`)
    })
  }
}
