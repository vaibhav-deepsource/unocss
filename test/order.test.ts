import { createGenerator } from '@unocss/core'
import { describe, expect, test } from 'vitest'
import { variantMatcher } from '@unocss/preset-mini/utils'

describe('order', () => {
  test('static', async() => {
    const uno = createGenerator({
      rules: [
        ['foo', { name: 'bar1' }],
        ['foo', { name: 'bar2' }],
      ],
      presets: [],
    })
    const { css } = await uno.generate('foo')
    expect(css).toContain('bar2')
  })

  test('dynamic', async() => {
    const uno = createGenerator({
      rules: [
        [/^foo$/, () => ({ name: 'bar1' })],
        [/^foo$/, () => ({ name: 'bar2' })],
      ],
      presets: [],
    })
    const { css } = await uno.generate('foo')
    expect(css).toContain('bar2')
  })

  test('variant', async() => {
    const uno = createGenerator({
      rules: [
        [/^foo-.$/, ([m]) => ({ name: m })],
      ],
      presets: [],
      variants: [
        variantMatcher('light', input => `.light $$ ${input}`),
        variantMatcher('group', input => `.group ${input}`),
        (v) => {
          const match = variantMatcher('dark', input => `.dark $$ ${input}`)(v)
          if (match) {
            return {
              ...match,
              order: 1,
            }
          }
        },
      ],
    })
    const code = [
      'light:group:foo-1',
      'group:light:foo-2',
      'dark:group:foo-3',
      'group:dark:foo-4',
    ].join(' ')
    const { css } = await uno.generate(code)
    const { css: css2 } = await uno.generate(code)
    expect(css).toMatchSnapshot()
    expect(css).toEqual(css2)
  })
})
