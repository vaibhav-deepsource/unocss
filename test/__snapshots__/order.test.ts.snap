// Vitest Snapshot v1

exports[`order > variant 1`] = `
"/* layer: default */
.dark .group .dark\\\\:group\\\\:foo-3{name:foo-3;}
.dark .group .group\\\\:dark\\\\:foo-4{name:foo-4;}
.group .light .light\\\\:group\\\\:foo-1{name:foo-1;}
.light .group .group\\\\:light\\\\:foo-2{name:foo-2;}"
`;
